dm/dm/master/scheduler/scheduler.go

```diff
@@ -2362,51 +2260,49 @@ func (s *Scheduler) tryBoundForSource(source string) (bool, error) {
 				continue
 			}
 			// the worker is not Offline
-			if w.Stage() == WorkerRelay {
+			if w.Stage() != WorkerOffline && len(w.Bounds()) < boundSourcesNum {
 				worker = w
+				boundSourcesNum = len(w.Bounds())
 				s.logger.Info("found relay worker when source bound",
 					zap.String("worker", workerName),
 					zap.String("source", source))
-				break
 			}
 		}
 	}
-	// then a history worker for this source...
 	if worker == nil {
-		for _, bound := range s.lastBound {
-			workerName := bound.Worker
-			if bound.Source == source {
-				w, ok := s.workers[workerName]
-				if !ok {
-					// a not found worker
-					continue
-				}
-				if w.Stage() == WorkerFree {
-					worker = w
-					s.logger.Info("found history worker when source bound",
-						zap.String("worker", workerName),
-						zap.String("source", source))
-					break
-				}
+		// then a history worker for this source...
+		if bound, ok := s.lastBound[source]; ok {
+			w, ok := s.workers[bound.Worker]
+			if !ok {
+				// a not found worker, should not happen
+				s.logger.DPanic("worker instance not found for history worker", zap.String("worker", bound.Worker))
+				return false, nil
+			}
+
+			if w.Stage() != WorkerOffline {
+				worker = w
+				s.logger.Info("found history worker when source bound",
+					zap.String("worker", workerName),
+					zap.String("source", source))
 			}
 		}
 	}
 
-	// and then a random Free worker.
 	if worker == nil {
+		// and then a Free worker with lease sources.
```

Comment:
```suggestion
		// and then a online worker with lease sources.
```

Reply:
addressed in https://github.com/pingcap/tiflow/pull/5077/commits/4f9d9f11200d41ebb97fe5a479896dd1080b2b9e